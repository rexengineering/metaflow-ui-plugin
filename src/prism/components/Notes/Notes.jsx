import React, { useCallback, useState } from "react";
import {
  Typography,
  Button,
  Card,
  CardContent,
  CardActions,
  CardHeader,
  IconButton,
  makeStyles,
} from "@material-ui/core";
import { Formik } from "formik";
import { object as yupObject } from "yup";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt } from "@fortawesome/pro-solid-svg-icons";
import TaskField from "../fields/TaskField";
import { TEXT, validationSchemaMapping } from "../../constants/taskTypes";

const useStyles = makeStyles((theme) => ({
  card: {
    marginBottom: theme.spacing(1),
  },
  notesContainer: {
    overflowY: "auto",
  },
  userNote: {
    whiteSpace: "pre-wrap",
  },
}));

function Notes() {
  const [notes, setNotes] = useState([]);
  const classes = useStyles();
  const initialValues = { newNote: "" };
  const validationSchema = yupObject({
    newNote: validationSchemaMapping[TEXT],
  });

  const onTextAreaKeyDown = useCallback((event, handleSubmit) => {
    if (event?.keyCode === 13 && !event?.shiftKey) {
      event?.preventDefault();
      event?.stopPropagation();
      handleSubmit();
    }
  }, []);

  const addNote = useCallback(
    ({ newNote }, { resetForm }) => {
      setNotes((currentNotes) => [
        ...currentNotes,
        { id: Math.random(), message: newNote, date: new Date() },
      ]);
      resetForm();
    },
    [setNotes]
  );

  const removeNote = useCallback((id) => {
    setNotes((currentNotes) => currentNotes.filter((note) => note.id !== id));
  }, []);

  return (
    <>
      <Typography variant="h5" align="center">
        Notes
      </Typography>
      <div className={classes.notesContainer}>
        {Array.isArray(notes) &&
          notes.map(({ id, message, date }) => (
            <Card key={id} className={classes.card} variant="outlined">
              <CardHeader
                action={
                  <IconButton onClick={() => removeNote(id)}>
                    <FontAwesomeIcon icon={faTrashAlt} />
                  </IconButton>
                }
                title={date.toLocaleTimeString()}
              />
              <CardContent>
                <Typography
                  className={classes.userNote}
                  data-testid="note-message"
                >
                  {message}
                </Typography>
              </CardContent>
            </Card>
          ))}
      </div>
      <Formik
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={addNote}
      >
        {({ handleSubmit }) => (
          <Card onSubmit={handleSubmit} component="form">
            <CardContent>
              <TaskField
                id="newNote"
                type={TEXT}
                multiline
                fullWidth
                label="New Note"
                onKeyDown={(event) => onTextAreaKeyDown(event, handleSubmit)}
              />
            </CardContent>
            <CardActions>
              <Button type="submit">Add</Button>
            </CardActions>
          </Card>
        )}
      </Formik>
    </>
  );
}

export default Notes;
