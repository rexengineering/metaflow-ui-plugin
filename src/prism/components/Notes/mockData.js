const mockData = {
  notes: [
    {
      id: "note1",
      message: "This is the message for note 1.",
    },
    {
      id: "note2",
      message: "This is the message for note 2.",
    },
  ],
};

export default mockData;
