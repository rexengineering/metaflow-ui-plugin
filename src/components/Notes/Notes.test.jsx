import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Notes from ".";

describe("<Notes />", () => {
  it("should render", () => {
    const { asFragment } = render(<Notes />);
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<Notes />);
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should add notes", async () => {
    render(<Notes />);

    const notes = ["my note", "my note 2", "my note 3"];
    const addButton = screen.getByRole("button");
    const input = screen.getByRole("textbox");

    for (let i = 0; i < notes.length; i++) {
      userEvent.type(input, notes[i]);
      // Disabled as we want these promises to be fired
      // sequentially. Testing library does not support
      // overlapping waitFor calls.
      // eslint-disable-next-line no-await-in-loop
      await waitFor(async () => {
        userEvent.click(addButton);
      });
    }

    const noteElem = screen.getAllByTestId("note-message");
    expect(noteElem).toHaveLength(3);

    notes.forEach((note, index) => {
      expect(noteElem[index]).toHaveTextContent(note);
    });
  });

  it("should clear the input when adding a note", async () => {
    render(<Notes />);

    const note = "my note";
    const addButton = screen.getByRole("button");
    const input = screen.getByRole("textbox");

    userEvent.type(input, note);
    await waitFor(async () => {
      userEvent.click(addButton);
    });

    expect(input).toHaveValue("");
  });

  it("should delete a note", async () => {
    render(<Notes />);

    const notes = ["my note", "my note 2", "my note 3"];
    const addButton = screen.getByRole("button");
    const input = screen.getByRole("textbox");

    for (let i = 0; i < notes.length; i++) {
      userEvent.type(input, notes[i]);
      // Disabled as we want these promises to be fired
      // sequentially. Testing library does not support
      // overlapping waitFor calls.
      // eslint-disable-next-line no-await-in-loop
      await waitFor(async () => {
        userEvent.click(addButton);
      });
    }

    const buttons = screen.getAllByRole("button");
    await waitFor(() => userEvent.click(buttons[1]))

    const noteElem = screen.getAllByTestId("note-message");
    expect(noteElem).toHaveLength(2);
    expect(noteElem[0]).toHaveTextContent(notes[0]);
    expect(noteElem[1]).toHaveTextContent(notes[2]);
  });
});
