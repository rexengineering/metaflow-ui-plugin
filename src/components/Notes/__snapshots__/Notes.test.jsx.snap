// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notes /> should match snapshot 1`] = `
<DocumentFragment>
  <h5
    class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter"
  >
    Notes
  </h5>
  <div
    class="makeStyles-notesContainer-5"
  />
  <form
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
          data-shrink="false"
        >
          New Note
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiFilledInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            name="newNote"
            rows="1"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Add
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</DocumentFragment>
`;
